### PR TITLE
#137: Added PayloadSender

### LINKED ISSUE #137: Add PayloadSender
# Background

For reliable send and receive, we need a PayloadSender which sends the Payload repeatedly until it got the acknowledgment. Furthermore, we need to be able to add sequence numbers, to restore the order at receiver.

### PR DESCRIPTION
**Pull request by tkilias**
Fixes #137

### CODE REVIEW

File: exasol_advanced_analytics_framework/udf_communication/peer_communicator/payload_message_sender.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+        self._frames = frames
+        self._message = message
+        self._finished = False
+        self._send_payload()
+
+    def _send_payload(self):
+        self._sender.send_multipart(self._frames)
+
+    def try_send(self):
+        should_we_send_abort = self._should_we_send_abort()
```
Comment (tomuben) [inline exasol_advanced_analytics_framework/udf_communication/peer_communicator/payload_message_sender.py:30]: maybe better `shoud_we_abort_send`:
```suggestion
        should_we_abort_send = self._should_we_abort_send()
```

File: exasol_advanced_analytics_framework/udf_communication/peer_communicator/payload_message_sender.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+            return
+        should_we_send_payload = self._should_we_send_payload()
+        if should_we_send_payload:
+            self._send_payload()
+            self._retry_timer.reset_timer()
+
+    def stop(self):
+        self._finished = True
+
+    def _should_we_send_abort(self):
```
Comment (tomuben) [inline exasol_advanced_analytics_framework/udf_communication/peer_communicator/payload_message_sender.py:43]:
```suggestion
    def _should_we_abort_send(self):
```

File: exasol_advanced_analytics_framework/udf_communication/peer_communicator/payload_message_sender.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        is_time = self._abort_timer.is_time()
+        is_enabled = not self._finished
+        return is_time and is_enabled
+
+    def _should_we_send_payload(self):
+        is_time = self._retry_timer.is_time()
+        is_enabled = not self._finished
+        return is_time and is_enabled
```
Comment (tomuben) [inline exasol_advanced_analytics_framework/udf_communication/peer_communicator/payload_message_sender.py:51]: Doesn't that mean that the initial send will happen only after the retry_timer fired for the first time?

Comment (tomuben) [reply exasol_advanced_analytics_framework/udf_communication/peer_communicator/payload_message_sender.py:51]: Ah got it, you do the initial send already in the constructor

File: exasol_advanced_analytics_framework/udf_communication/peer_communicator/payload_message_sender.py [hunk 4]
```diff
@@ -27,7 +27,7 @@ def _send_payload(self):
         self._sender.send_multipart(self._frames)
 
     def try_send(self):
-        should_we_send_abort = self._should_we_send_abort()
+        should_we_abort = self._should_we_abort()
         if should_we_send_abort:
```
Comment (tomuben) [inline exasol_advanced_analytics_framework/udf_communication/peer_communicator/payload_message_sender.py:31]: sorry, need to apply here, too

File: exasol_advanced_analytics_framework/udf_communication/peer_communicator/payload_message_sender.py [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+        self._message = message
+        self._finished = False
+        self._send_payload()
+
+    def _send_payload(self):
+        self._sender.send_multipart(self._frames)
+
+    def try_send(self):
+        should_we_send_abort = self._should_we_abort()
+        if should_we_send_abort:
```
Comment (tomuben) [inline exasol_advanced_analytics_framework/udf_communication/peer_communicator/payload_message_sender.py:31]:
```suggestion
        if should_we_abort:
```

File: exasol_advanced_analytics_framework/udf_communication/peer_communicator/background_listener_thread.py [hunk 6]
```diff
@@ -280,7 +280,7 @@ def is_register_peer_message_allowed_as_listener_message(self) -> bool:
                and self._config.forward_register_peer_config.is_enabled
 
     def _handle_payload_message(self, message: messages.Payload, frames: List[Frame]):
-        peer = Peer(connection_info=message.source)
+        peer = connection_info = message.source
```
Comment (tomuben) [inline exasol_advanced_analytics_framework/udf_communication/peer_communicator/background_listener_thread.py:283]: Is this correct?

Comment (tkilias) [inline exasol_advanced_analytics_framework/udf_communication/peer_communicator/background_listener_thread.py:283]:
```suggestion
        peer = message.source
```

Comment (tkilias) [reply exasol_advanced_analytics_framework/udf_communication/peer_communicator/background_listener_thread.py:283]: Not correct, but also not wrong

**Review by tomuben — APPROVED** [review]

**Pull request merged**